sdks/java/io/hadoop-format/src/main/java/org/apache/beam/sdk/io/hadoop/format/HadoopFormatIO.java

```diff
@@ -0,0 +1,1182 @@
+  private static class ConfigurationCoder extends AtomicCoder<Configuration> {
+
+    @Override
+    public void encode(Configuration value, OutputStream outStream) throws IOException {
+      DataOutputStream dataOutputStream = new DataOutputStream(outStream);
+      value.write(dataOutputStream);
+      dataOutputStream.flush();
+    }
+
+    @Override
+    public Configuration decode(InputStream inStream) throws IOException {
+      DataInputStream dataInputStream = new DataInputStream(inStream);
+      Configuration config = new Configuration(false);
+      config.readFields(dataInputStream);
+
+      return config;
+    }
+  }
+
+  /**
+   * DoFn with following responsibilities:
+   *
+   * <ul>
+   *   <li>Validates configuration - checks whether all required properties are set.
+   *   <li>Validates types of input PCollection elements.
+   *   <li>Setups start of the {@link OutputFormat} job for given window.
+   * </ul>
+   *
+   * <p>Logic of the setup job is called only for the very first element of the instance. All other
+   * elements are directly sent to next processing.
+   */
+  private static class SetupJobFn<KeyT, ValueT> extends DoFn<KV<KeyT, ValueT>, KV<KeyT, ValueT>> {
+
+    private ExternalSynchronization externalSynchronization;
+    private PCollectionView<Configuration> configView;
+    private TypeDescriptor<KV<KeyT, ValueT>> inputTypeDescriptor;
+    private boolean isSetupJobAttempted;
+
+    SetupJobFn(
+        ExternalSynchronization externalSynchronization,
+        PCollectionView<Configuration> configView,
+        TypeDescriptor<KV<KeyT, ValueT>> inputTypeDescriptor) {
+      this.externalSynchronization = externalSynchronization;
+      this.configView = configView;
+      this.inputTypeDescriptor = inputTypeDescriptor;
+    }
+
+    @Setup
+    public void setup() {
+      isSetupJobAttempted = false;
+    }
+
+    @DoFn.ProcessElement
+    public void processElement(
+        @DoFn.Element KV<KeyT, ValueT> element,
+        OutputReceiver<KV<KeyT, ValueT>> receiver,
+        BoundedWindow window,
+        ProcessContext c) {
+
+      receiver.output(element);
+
+      if (isSetupJobAttempted) {
+        // setup of job was already attempted
+        return;
+      }
+
+      Configuration conf = c.sideInput(configView);
+
+      // validate configuration and input
+      // must be done first, because in all later operations are required assumptions from
+      // validation
+      validateConfiguration(conf);
+      validateInputData(conf);
+
+      boolean isJobLockAcquired = externalSynchronization.tryAcquireJobLock(conf);
+      isSetupJobAttempted = true;
+
+      if (!isJobLockAcquired) {
+        // some parallel execution acquired task
+        return;
+      }
+
+      try {
+        // setup job
+        JobID jobId = HadoopFormats.getJobId(conf);
+
+        trySetupJob(jobId, conf, window);
+
+      } catch (Exception e) {
+        //        externalSynchronization.releaseJobIdLock(conf);
```

Comment:
Presumably commented out, to block other running but when will it be cleaned up?

Reply:
Deleted, thank you.

Reply:
My question was actually when with the lock file be cleaned up?

Reply:
Oh sorry for misunderstanding.

If you think "when in normal flow", then it is done in [HadoopFormatIO.CommitJobFn#cleanupJob](https://github.com/apache/beam/blob/4193ac965b7c7cf09d195411a5e28ac1e697d11f/sdks/java/io/hadoop-format/src/main/java/org/apache/beam/sdk/io/hadoop/format/HadoopFormatIO.java#L908)

And regarding the exception catch in mentioned code. I think that it is not necessary there now to have it there because this block is guarded by lock. So any unexpected exception should not be thrown. On the other hand it could be nice debugging point when somebody will write new `ExternalSynchronization` which does not work properly.

Did I answer your question now?